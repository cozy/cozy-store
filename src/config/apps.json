{
  "notRemovableApps": ["drive"],
  "notDisplayedApps": ["collect", "home", "settings", "store", "dataproxy"]
}
