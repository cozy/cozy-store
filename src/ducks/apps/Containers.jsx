import DiscoverComponent from 'ducks/apps/components/Discover'
import MyApplicationsComponent from 'ducks/apps/components/MyApplications'
import SidebarCategoriesComponent from 'ducks/apps/components/SidebarCategories'
import { connect } from 'react-redux'

import { translate } from 'cozy-ui/transpiled/react/I18n'

import { getInstalledApps, getRegistryApps } from './index'

const mapStateToProps = state => ({
  apps: getRegistryApps(state),
  installedApps: getInstalledApps(state),
  isFetching: state.apps.isFetching,
  isAppFetching: state.apps.isAppFetching,
  isInstalling: state.apps.isInstalling,
  isUninstalling: state.apps.isUninstalling,
  actionError: state.apps.actionError,
  fetchError: state.apps.fetchError
})

export const Discover = translate()(connect(mapStateToProps)(DiscoverComponent))

export const MyApplications = translate()(
  connect(mapStateToProps)(MyApplicationsComponent)
)

export const SidebarCategories = connect(
  state => ({
    apps: getRegistryApps(state),
    installedApps: getInstalledApps(state)
  }),
  null
)(SidebarCategoriesComponent)
