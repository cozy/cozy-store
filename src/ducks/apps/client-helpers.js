import { APP_TYPE } from './constants'

/**
 * Fetch at most 200 apps from the channel
 *
 * Optional filter can be given
 */
export const fetchAppsFromChannel = (client, channel, filter) => {
  let filterParam = ''
  // TODO do it with URL or queryParams
  if (filter) filterParam = `&filter[type]=${filter}`
  return client.stackClient.fetchJSON(
    'GET',
    `/registry?limit=300&versionsChannel=${channel}&latestChannelVersion=${channel}${filterParam}`
  )
}

/**
 * Fetch user apps or konnectors.
 * Adds type property in attributes from stack
 */
export const fetchUserApps = async (client, type) => {
  const route = type === APP_TYPE.KONNECTOR ? '/konnectors/' : '/apps/'
  const { data } = await client.stackClient.fetchJSON('GET', route)
  return data.map(x => {
    x.attributes.type = type
    return x
  })
}

export const fetchAppOrKonnector = (client, doctype, slug) => {
  // FIXME: hack to handle node type from stack for the konnectors
  const route =
    doctype === APP_TYPE.KONNECTOR || doctype === 'node' ? 'konnectors' : 'apps'
  return client.stackClient
    .fetchJSON('GET', `/${route}/${slug}`)
    .then(x => x.data)
}
