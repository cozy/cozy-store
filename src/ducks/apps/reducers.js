'use strict'

import { extend as extendI18n } from 'cozy-ui/react/I18n'
import { combineReducers } from 'redux'

// initial loading
export const LOADING_APP = 'LOADING_APP'
export const LOADING_APP_INTENT = 'LOADING_APP_INTENT'

export const FETCH_APPS = 'FETCH_APPS'
export const FETCH_APPS_SUCCESS = 'FETCH_APPS_SUCCESS'
export const FETCH_APPS_FAILURE = 'FETCH_APPS_FAILURE'

export const FETCH_APP = 'FETCH_APP'
export const FETCH_APP_SUCCESS = 'FETCH_APP_SUCCESS'
export const FETCH_APP_FAILURE = 'FETCH_APP_FAILURE'

// store an app state, to handle app version switch cancelling
export const RESTORE_APP = 'RESTORE_APP'
export const SAVE_APP = 'SAVE_APP'

export const FETCH_REGISTRY_APPS_SUCCESS = 'FETCH_REGISTRY_APPS_SUCCESS'

export const UNINSTALL_APP = 'UNINSTALL_APP'
export const UNINSTALL_APP_SUCCESS = 'UNINSTALL_APP_SUCCESS'
export const UNINSTALL_APP_FAILURE = 'UNINSTALL_APP_FAILURE'

export const INSTALL_APP = 'INSTALL_APP'
export const INSTALL_APP_SUCCESS = 'INSTALL_APP_SUCCESS'
export const INSTALL_APP_FAILURE = 'INSTALL_APP_FAILURE'

export function _sortAlphabetically(array, property) {
  return array.sort((a, b) => a[property] > b[property])
}

export function _isValidApp(app) {
  return !!app && !!app.slug && !!app.source && !!app.version
}

export function _consolidateApps(stateApps, newAppsInfos, lang) {
  const apps = new Map()
  stateApps.forEach(app => apps.set(app.slug, app))
  newAppsInfos.forEach(app => {
    const appFromState = apps.get(app.slug)
    // handle maintenance locales
    let appLocales = app.locales
    if (appLocales && appFromState && appFromState.locales) {
      for (let locale in appFromState.locales) {
        appLocales[locale] = Object.assign(
          {},
          appFromState.locales[locale],
          app.locales[locale]
        )
      }
    }
    if (appLocales && appLocales[lang]) {
      // access app locales from 'apps.slug.[...]'
      extendI18n({ apps: { [app.slug]: appLocales[lang] } })
    }
    if (appFromState) {
      apps.set(
        app.slug,
        Object.assign({}, appFromState, app, {
          locales: appLocales
        })
      )
    } else {
      apps.set(app.slug, app)
    }
  })
  return Array.from(apps.values()).filter(app => app)
}

export const list = (state = [], action = {}) => {
  switch (action.type) {
    case RESTORE_APP:
      if (_isValidApp(action.app)) {
        return _sortAlphabetically(
          _consolidateApps(
            // we completely replace it, so we remove the current from state
            state.filter(a => a.slug !== action.app.slug),
            [action.app],
            action.lang
          ),
          'slug'
        )
      } else {
        console.warn('Failed attempt to restore a saved app state.')
        return state
      }
    case FETCH_APP_SUCCESS:
      return _sortAlphabetically(
        _consolidateApps(state, [action.app], action.lang),
        'slug'
      )
    case FETCH_REGISTRY_APPS_SUCCESS:
    case FETCH_APPS_SUCCESS:
      return _sortAlphabetically(
        _consolidateApps(state, action.apps, action.lang),
        'slug'
      )
    case UNINSTALL_APP_SUCCESS:
      return state.map(app =>
        app.slug === action.slug ? { ...app, installed: false } : app
      )
    case INSTALL_APP_SUCCESS: {
      return _sortAlphabetically(
        state.map(app => {
          if (app.slug === action.installedApp.slug) {
            const nextApp = { ...app, ...action.installedApp, installed: true }
            // the available update is now installed
            if (nextApp.availableVersion) delete nextApp.availableVersion
            return nextApp
          }
          return app
        }),
        'slug'
      )
    }
    default:
      return state
  }
}

export const savedApp = (state = null, action = {}) => {
  switch (action.type) {
    case SAVE_APP:
      return _isValidApp(action.app) ? action.app : state
    case RESTORE_APP:
    case INSTALL_APP_SUCCESS: // after install we clean the previously saved app
      return null
    default:
      return state
  }
}

export const isFetching = (state = false, action = {}) => {
  switch (action.type) {
    case LOADING_APP:
    case FETCH_APPS:
      return true
    case FETCH_APPS_SUCCESS:
    case FETCH_APPS_FAILURE:
      return false
    default:
      return state
  }
}

export const isAppFetching = (state = false, action = {}) => {
  switch (action.type) {
    case LOADING_APP_INTENT:
    case FETCH_APP:
      return true
    case RESTORE_APP: // restoring the saved app cancels the app fetching
      return _isValidApp(action.app) ? false : state
    case FETCH_APP_SUCCESS:
    case FETCH_APP_FAILURE:
      return false
    default:
      return state
  }
}

export const isInstalling = (state = false, action = {}) => {
  switch (action.type) {
    case INSTALL_APP:
      return action.slug
    case INSTALL_APP_SUCCESS:
    case INSTALL_APP_FAILURE:
      return false
    default:
      return state
  }
}

export const isUninstalling = (state = false, action = {}) => {
  switch (action.type) {
    case UNINSTALL_APP:
      return true
    case UNINSTALL_APP_SUCCESS:
    case UNINSTALL_APP_FAILURE:
      return false
    default:
      return state
  }
}

export const actionError = (state = null, action = {}) => {
  switch (action.type) {
    case UNINSTALL_APP_FAILURE:
    case INSTALL_APP_FAILURE:
      return action.error
    case UNINSTALL_APP_SUCCESS:
    case INSTALL_APP_SUCCESS:
      return null
    default:
      return state
  }
}

export const fetchError = (state = null, action = {}) => {
  switch (action.type) {
    case FETCH_APPS_FAILURE:
    case FETCH_APP_FAILURE:
      return action.error
    case FETCH_APPS_SUCCESS:
    case FETCH_APP_SUCCESS:
      return null
    default:
      return state
  }
}

export const appsReducers = combineReducers({
  list,
  actionError,
  fetchError,
  isFetching,
  isAppFetching,
  isInstalling,
  isUninstalling,
  savedApp
})
