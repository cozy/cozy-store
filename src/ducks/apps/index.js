/* eslint-env browser */
/* global cozy */

import config from 'config/apps'
import CONSTANTS from 'config/constants'
import AUTHORIZED_CATEGORIES from 'config/categories'
import { NotUninstallableAppException } from '../../lib/exceptions'
import realtime from 'cozy-realtime'

export * from './selectors'
export { appsReducers } from './reducers'
import {
  LOADING_APP,
  LOADING_APP_INTENT,
  FETCH_APPS,
  FETCH_APPS_SUCCESS,
  FETCH_APPS_FAILURE,
  FETCH_APP,
  FETCH_APP_SUCCESS,
  FETCH_APP_FAILURE,
  FETCH_REGISTRY_APPS_SUCCESS,
  UNINSTALL_APP,
  UNINSTALL_APP_SUCCESS,
  UNINSTALL_APP_FAILURE,
  INSTALL_APP,
  INSTALL_APP_SUCCESS,
  INSTALL_APP_FAILURE,
  RESTORE_APP,
  SAVE_APP
} from './reducers'

const APP_STATE = {
  READY: 'ready',
  INSTALLING: 'installing',
  ERRORED: 'errored'
}

export const APP_TYPE = {
  KONNECTOR: 'konnector',
  WEBAPP: 'webapp'
}

export const REGISTRY_CHANNELS = {
  DEV: 'dev',
  BETA: 'beta',
  STABLE: 'stable'
}

const APPS_DOCTYPE = 'io.cozy.apps'
const KONNECTORS_DOCTYPE = 'io.cozy.konnectors'
const TERMS_DOCTYPE = 'io.cozy.terms'

const DEFAULT_CHANNEL = CONSTANTS.default.registry.channel

/* Only for the icon fetching */
let dataset
const getDataset = () => {
  if (dataset) return dataset
  const root = document.querySelector('[role=application]')
  dataset = root && root.dataset
  return dataset
}
/* Only for the icon fetching */

export const getAppIconProps = () => ({
  domain: getDataset() && getDataset().cozyDomain,
  secure: window.location.protocol === 'https:'
})

export function _sanitizeManifest(app) {
  // FIXME retro-compatibility for old formatted manifest
  const sanitized = Object.assign({}, app)
  if (!app.categories && app.category && typeof app.category === 'string') {
    sanitized.categories = [app.category]
    delete sanitized.category
  }
  if (typeof app.name === 'object') sanitized.name = app.name.en
  // FIXME use camelCase from cozy-stack
  if (app.available_version) {
    sanitized.availableVersion = app.available_version
    delete sanitized.available_version
  }
  // remove incomplete or empty terms
  const hasValidTerms =
    !!sanitized.terms &&
    !!sanitized.terms.id &&
    !!sanitized.terms.url &&
    !!sanitized.terms.version
  if (sanitized.terms && !hasValidTerms) delete sanitized.terms
  // remove incomplete or empty partnership
  const hasValidPartnership =
    !!sanitized.partnership && !!sanitized.partnership.description
  if (sanitized.partnership && !hasValidPartnership)
    delete sanitized.partnership
  return sanitized
}

// check authorized categories and add default 'others'
export function _sanitizeCategories(categoriesList) {
  if (!categoriesList) return ['others']
  const filteredList = categoriesList.filter(c =>
    AUTHORIZED_CATEGORIES.includes(c)
  )
  if (!filteredList.length) return ['others']
  return filteredList
}

let contextCache
export function getContext() {
  return contextCache
    ? Promise.resolve(contextCache)
    : cozy.client
        .fetchJSON('GET', '/settings/context')
        .then(context => {
          contextCache = context
          return context
        })
        .catch(error => {
          if (error.status && error.status === 404) {
            contextCache = {}
            return contextCache
          }
          return {}
        })
}

export function _getRegistryAssetsLinks(manifest, appVersion) {
  if (!appVersion && manifest) appVersion = manifest.version
  if (!appVersion) return {}
  const screenshotsLinks =
    manifest.screenshots &&
    manifest.screenshots.map(name => {
      let fileName = name
      if (fileName[0] === '/') fileName = fileName.slice(1)
      return `${cozy.client._url}/registry/${
        manifest.slug
      }/${appVersion}/screenshots/${fileName}`
    })
  const iconLink =
    manifest.slug && `/registry/${manifest.slug}/${appVersion}/icon`
  const partnershipIconLink =
    manifest.slug &&
    manifest.partnership &&
    manifest.partnership.icon &&
    `${cozy.client._url}/registry/${
      manifest.slug
    }/${appVersion}/partnership_icon`
  return {
    screenshotsLinks,
    iconLink,
    partnershipIconLink
  }
}

export async function getFormattedInstalledApp(response) {
  const appAttributes = _sanitizeManifest(response.attributes)

  const openingLink = response.links.related
  const { screenshotsLinks, partnershipIconLink } = _getRegistryAssetsLinks(
    appAttributes,
    appAttributes.version
  )
  const partnership =
    !!appAttributes.partnership &&
    Object.assign({}, appAttributes.partnership, {
      ...(partnershipIconLink ? { icon: partnershipIconLink } : {})
    })
  return Object.assign({}, appAttributes, {
    _id: response.id || response._id,
    categories: _sanitizeCategories(appAttributes.categories),
    installed: true,
    related: openingLink,
    links: response.links,
    // Add partnership property only if it exists
    ...(partnership ? { partnership } : {}),
    // add screenshotsLinks property only if it exists
    ...(screenshotsLinks ? { screenshots: screenshotsLinks } : {}),
    uninstallable: !config.notRemovableApps.includes(appAttributes.slug)
  })
}

// only on the app initialisation
export function initApp(lang) {
  return async dispatch => {
    dispatch({ type: LOADING_APP })
    await dispatch(initializeRealtime())
    return dispatch(fetchApps(lang))
  }
}

// only on the app install intent initialisation
export function initAppIntent(lang, slug) {
  return async dispatch => {
    dispatch({ type: LOADING_APP_INTENT })
    await dispatch(initializeRealtime())
    return await dispatch(fetchLatestApp(lang, slug))
  }
}

function onAppUpdate(appResponse) {
  return async dispatch => {
    if (appResponse.state === APP_STATE.ERRORED) {
      const err = new Error('Error when installing the application')
      dispatch({ type: INSTALL_APP_FAILURE, error: err })
      throw err
    }
    if (appResponse.state === APP_STATE.READY) {
      // FIXME: hack to handle node type from stack for the konnectors
      const route =
        appResponse.type === APP_TYPE.KONNECTOR || appResponse.type === 'node'
          ? 'konnectors'
          : 'apps'
      const appFromStack = await cozy.client.fetchJSON(
        'GET',
        `/${route}/${appResponse.slug}`
      )
      return getFormattedInstalledApp(appFromStack).then(installedApp => {
        return dispatch({ type: INSTALL_APP_SUCCESS, installedApp })
      })
    }
  }
}

function onAppDelete(appResponse) {
  return async dispatch => {
    if (appResponse.state === APP_STATE.ERRORED) {
      const err = new Error('Error when installing the application')
      dispatch({ type: UNINSTALL_APP_FAILURE, error: err })
      throw err
    }
    const { slug } = appResponse
    return dispatch({ type: UNINSTALL_APP_SUCCESS, slug })
  }
}

function initializeRealtime() {
  const config = {
    token: cozy.client._token.token,
    // cozy-realtime expect an URL with an https protocol,
    // see https://github.com/cozy/cozy-libs/blob/master/packages/realtime/src/index.js#L52
    url: `${window.location.protocol}${cozy.client._url}`
  }
  return async dispatch => {
    realtime
      .subscribeAll(config, APPS_DOCTYPE)
      .then(subscription => {
        // HACK: the push CREATE at fisrt install
        subscription.onCreate(app => dispatch(onAppUpdate(app)))
        subscription.onUpdate(app => dispatch(onAppUpdate(app)))
        subscription.onDelete(app => dispatch(onAppDelete(app)))
      })
      .catch(error => {
        // eslint-disable-next-line no-console
        console.warn(`Cannot initialize realtime for apps: ${error.message}`)
      })

    realtime
      .subscribeAll(config, KONNECTORS_DOCTYPE)
      .then(subscription => {
        // HACK: the push CREATE at fisrt install
        subscription.onCreate(app => dispatch(onAppUpdate(app)))
        subscription.onUpdate(app => dispatch(onAppUpdate(app)))
        subscription.onDelete(app => dispatch(onAppDelete(app)))
      })
      .catch(error => {
        // eslint-disable-next-line no-console
        console.warn(
          `Cannot initialize realtime for konnectors: ${error.message}`
        )
      })
  }
}

async function _getInstalledInfos(app) {
  try {
    let installedApp = await cozy.client.fetchJSON(
      'GET',
      `/${app.type === APP_TYPE.WEBAPP ? 'apps' : 'konnectors'}/${app.slug}`
    )
    return !!installedApp
  } catch (e) {
    return false
  }
}

/* Restore a previous saved app state into the apps list */
export function restoreAppIfSaved() {
  return async (dispatch, getState) => {
    const savedApp = getState().apps.savedApp
    if (savedApp && getState().apps.isInstalling !== savedApp.slug)
      dispatch({ type: RESTORE_APP, app: savedApp })
  }
}

export function fetchLatestApp(
  lang,
  slug,
  channel = DEFAULT_CHANNEL,
  appToSave = null // if provided, we store it into savedApp state
) {
  return async (dispatch, getState) => {
    if (appToSave) dispatch({ type: SAVE_APP, app: appToSave })
    dispatch({ type: FETCH_APP })
    let app = getState().apps.list.find(a => a.slug === slug)
    try {
      app = await cozy.client.fetchJSON(
        'GET',
        `/registry/${slug}?latestChannelVersion=${channel}`
      )
    } catch (err) {
      let errorMessage = `Error while getting the application with slug: ${slug}`
      if (err.status === 404) {
        errorMessage = `Application ${slug} not found in the registry.`
      }
      return dispatch({
        type: FETCH_APP_FAILURE,
        error: new Error(errorMessage)
      })
    }
    try {
      const formattedApp = await getFormattedRegistryApp(app, channel)
      formattedApp.installed = await _getInstalledInfos(app)
      // dispatch the app only if we are still in fetching status
      // (meant that the fetch has not been cancelled by a RESTORE_APP action)
      if (getState().apps.isAppFetching) {
        dispatch({
          type: FETCH_APP_SUCCESS,
          app: formattedApp,
          lang
        })
      }
      return formattedApp
    } catch (err) {
      if (err.status === 404) {
        // eslint-disable-next-line no-console
        console.warn(
          `No ${channel} version found for ${slug} from the registry.`
        )
      } else {
        // eslint-disable-next-line no-console
        console.warn(
          `Something went wrong when trying to fetch more informations about ${slug} from the registry on ${channel} channel. ${err}`
        )
      }
      dispatch({ type: FETCH_APP_FAILURE, error: err })
      throw err
    }
  }
}

export async function getFormattedRegistryApp(
  responseApp,
  channel = DEFAULT_CHANNEL
) {
  let version = responseApp.latest_version
  if (!version) {
    version = await cozy.client.fetchJSON(
      'GET',
      `/registry/${responseApp.slug}/${channel}/latest`
    )
  }
  const manifest = _sanitizeManifest(version.manifest)
  // source is only used by the stack when installed
  // so we removed it from the app manifest if present
  if (manifest.source) delete manifest.source

  // handle locales with maintenance status and messages
  let appLocales = manifest.locales
  let maintenance = null
  if (responseApp.maintenance_activated) {
    maintenance = Object.assign({}, responseApp.maintenance_options)
    if (maintenance.messages) {
      for (let lang in maintenance.messages) {
        if (appLocales[lang]) {
          appLocales[lang].maintenance = maintenance.messages[lang]
        }
      }
      delete maintenance.messages
    }
  }

  const versionFromRegistry = version.version
  const {
    screenshotsLinks,
    iconLink,
    partnershipIconLink
  } = _getRegistryAssetsLinks(manifest, versionFromRegistry)
  const partnership =
    !!manifest.partnership &&
    Object.assign({}, manifest.partnership, {
      ...(partnershipIconLink ? { icon: partnershipIconLink } : {})
    })
  return Object.assign(
    {},
    {
      versions: responseApp.versions
    },
    manifest,
    {
      version: versionFromRegistry,
      type: version.type,
      categories: _sanitizeCategories(manifest.categories),
      links: { icon: iconLink },
      uninstallable: !config.notRemovableApps.includes(manifest.slug),
      isInRegistry: true,
      // Add partnership property only if it exists
      ...(partnership ? { partnership } : {}),
      // handle maintenance status
      ...(maintenance ? { maintenance } : {}),
      // add screenshotsLinks property only if it exists
      ...(screenshotsLinks ? { screenshots: screenshotsLinks } : {}),
      // add installed value only if not already provided
      installed: responseApp.installed || false
    }
  )
}

export function fetchInstalledApps(lang, fetchingRegistry) {
  return async dispatch => {
    dispatch({ type: FETCH_APPS })
    try {
      // Start the HTTP requests as soon as possible
      let fetchingKonnectors = cozy.client.fetchJSON('GET', '/konnectors/')
      let fetchingWebApps = cozy.client.fetchJSON('GET', '/apps/')
      await fetchingRegistry
      let installedWebApps = await fetchingWebApps
      installedWebApps = installedWebApps.map(w => {
        // FIXME type konnector is missing from stack
        w.attributes.type = APP_TYPE.WEBAPP
        return w
      })
      // TODO throw error if collect is not installed
      const collectApp = installedWebApps.find(
        a => a.attributes.slug === 'collect'
      )
      const collectLink = collectApp && collectApp.links.related
      installedWebApps = installedWebApps.filter(
        app => !config.notDisplayedApps.includes(app.attributes.slug)
      )
      let installedKonnectors = await fetchingKonnectors
      installedKonnectors = installedKonnectors.map(k => {
        // FIXME type konnector is missing from stack
        k.attributes.type = APP_TYPE.KONNECTOR
        return k
      })
      installedKonnectors = installedKonnectors.filter(
        app => !config.notDisplayedApps.includes(app.attributes.slug)
      )
      const installedApps = installedWebApps.concat(installedKonnectors)
      Promise.all(
        installedApps.map(app => {
          return getFormattedInstalledApp(app, collectLink, false)
        })
      ).then(apps => {
        return dispatch({ type: FETCH_APPS_SUCCESS, apps, lang })
      })
    } catch (e) {
      dispatch({ type: FETCH_APPS_FAILURE, error: e })
      throw e
    }
  }
}

export function fetchRegistryApps(lang, channel = DEFAULT_CHANNEL) {
  return dispatch => {
    dispatch({ type: FETCH_APPS })
    return cozy.client
      .fetchJSON(
        'GET',
        `/registry?limit=200&versionsChannel=${channel}&latestChannelVersion=${channel}`
      )
      .then(response => {
        const apps = response.data
          .filter(app => !config.notDisplayedApps.includes(app.slug))
          .filter(app => app.versions[channel] && app.versions[channel].length) // only apps with versions available
        return Promise.all(
          apps.map(app => {
            // no latest_version means no version for this channel
            if (!app.latest_version) return false // skip
            return getFormattedRegistryApp(app).catch(err => {
              console.warn(
                `Something went wrong when trying to fetch more informations about ${
                  app.slug
                } from the registry on ${channel} channel. ${err}`
              )
              return false // skip
            })
          })
        ).then(apps => {
          return dispatch({
            type: FETCH_REGISTRY_APPS_SUCCESS,
            apps: apps.filter(a => a),
            lang
          })
        })
      })
      .catch(e => {
        dispatch({ type: FETCH_APPS_FAILURE, error: e })
        throw e
      })
  }
}

export function fetchApps(lang) {
  return async dispatch => {
    const fetchingRegistry = dispatch(fetchRegistryApps(lang))
    return dispatch(fetchInstalledApps(lang, fetchingRegistry))
  }
}

function extractJsonApiError(e) {
  try {
    const parsed = JSON.parse(e.message)
    return new Error(parsed.errors[0].detail || e.message)
  } catch (err) {
    return e
  }
}

export function uninstallApp(app) {
  const { slug, type } = app
  return dispatch => {
    if (
      config.notRemovableApps.includes(slug) ||
      config.notDisplayedApps.includes(slug)
    ) {
      const error = new NotUninstallableAppException()
      dispatch({ type: UNINSTALL_APP_FAILURE, error })
    }
    dispatch({ type: UNINSTALL_APP })
    // FIXME: hack to handle node type from stack for the konnectors
    const route =
      type === APP_TYPE.KONNECTOR || type === 'node' ? 'konnectors' : 'apps'
    return cozy.client.fetchJSON('DELETE', `/${route}/${slug}`).catch(e => {
      dispatch({ type: UNINSTALL_APP_FAILURE, error: extractJsonApiError(e) })
    })
  }
}

let termsIndexCache = null
async function _getOrCreateTermsIndex() {
  termsIndexCache = await cozy.client.data.defineIndex(TERMS_DOCTYPE, [
    'termsId',
    'version'
  ])
  return termsIndexCache
}

async function _saveAppTerms(terms) {
  const { id, ...termsAttributes } = terms
  // We use : as separator for <id>:<version>
  let savedTermsDocs = null
  try {
    savedTermsDocs = await cozy.client.data.query(
      await _getOrCreateTermsIndex(),
      {
        selector: {
          termsId: id,
          version: termsAttributes.version
        },
        limit: 1
      }
    )
  } catch (e) {
    throw e
  }
  if (savedTermsDocs && savedTermsDocs.length) {
    // we just update the url if this is the same id and same version
    // but the url changed
    const savedTerms = savedTermsDocs[0]
    if (
      savedTerms.termsId == id &&
      savedTerms.version == termsAttributes.version &&
      savedTerms.url != termsAttributes.url
    ) {
      await cozy.client.data.updateAttributes(TERMS_DOCTYPE, savedTerms._id, {
        url: termsAttributes.url
      })
    }
  } else {
    const termsToSave = Object.assign({}, termsAttributes, {
      termsId: id,
      accepted: true,
      acceptedAt: new Date()
    })
    await cozy.client.data.create(TERMS_DOCTYPE, termsToSave)
  }
}

export function installApp(app, source, isUpdate = false) {
  const { slug, type, terms } = app
  return async dispatch => {
    dispatch({ type: INSTALL_APP, slug })
    const handleError = e => {
      dispatch({ type: INSTALL_APP_FAILURE, error: e })
      throw e
    }
    const args = {}
    if (isUpdate) args.PermissionsAcked = isUpdate
    if (source) args.Source = source
    const queryString = Object.keys(args)
      .map(k => k + '=' + args[k])
      .join('&')
    if (terms) {
      try {
        await _saveAppTerms(terms)
      } catch (e) {
        handleError(e)
      }
    }
    const verb = isUpdate ? 'PUT' : 'POST'
    const route = type === APP_TYPE.KONNECTOR ? 'konnectors' : 'apps'
    try {
      await cozy.client.fetchJSON(verb, `/${route}/${slug}?${queryString}`)
    } catch (e) {
      handleError(e)
    }
  }
}

export function installAppFromRegistry(
  app,
  channel = DEFAULT_CHANNEL,
  isUpdate = false
) {
  return dispatch => {
    const source = `registry://${app.slug}/${channel}`
    return dispatch(installApp(app, source, isUpdate))
  }
}
