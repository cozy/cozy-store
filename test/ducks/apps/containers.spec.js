import '@testing-library/jest-dom'
import { render } from '@testing-library/react'
import { MyApplications, Discover } from 'ducks/apps/Containers'
import React from 'react'
import configureStore from 'redux-mock-store'

import AppLike from '../../AppLike'

const mockStore = configureStore()

describe('Apps Containers (connected components):', () => {
  const initialState = {
    apps: {
      list: [],
      isFetching: false,
      isInstalling: false,
      actionError: null,
      fetchError: null
    }
  }

  it('MyApplications should be rendered correctly at initial store', () => {
    const store = mockStore(initialState)
    const { container } = render(
      <AppLike>
        <MyApplications store={store} />
      </AppLike>
    )
    expect(container.querySelector('.sto-myapps')).toBeInTheDocument()
  })

  it('Discover should be rendered correctly at initial store', () => {
    const store = mockStore(initialState)
    const { container } = render(
      <AppLike>
        <Discover store={store} />
      </AppLike>
    )
    expect(container.querySelector('.sto-discover')).toBeInTheDocument()
  })
})
