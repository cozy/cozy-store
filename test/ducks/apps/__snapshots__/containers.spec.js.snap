// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Apps Containers (connected components): Discover should be rendered correctly at initial store 1`] = `
<withI18n(withBreakpoints(Discover))
  actionError={null}
  apps={Array []}
  dispatch={[Function]}
  fetchError={null}
  installedApps={Array []}
  isFetching={false}
  isInstalling={false}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;

exports[`Apps Containers (connected components): MyApplications should be rendered correctly at initial store 1`] = `
<withI18n(withBreakpoints(MyApplications))
  actionError={null}
  apps={Array []}
  dispatch={[Function]}
  fetchError={null}
  installedApps={Array []}
  isFetching={false}
  isInstalling={false}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;
